taichi/backends/cuda/cuda_profiler.cpp

```diff
@@ -1,19 +1,167 @@
 #include "taichi/backends/cuda/cuda_profiler.h"
 #include "taichi/backends/cuda/cuda_driver.h"
 #include "taichi/backends/cuda/cuda_context.h"
-#include "taichi/system/timeline.h"
 
 TLANG_NAMESPACE_BEGIN
 
 #if defined(TI_WITH_CUDA)
 
-std::string KernelProfilerCUDA::title() const {
-  return "CUDA Profiler";
+bool check_device_capability() {
+  void *device;
+  int cc_major;
+  CUDADriver::get_instance().device_get(&device, 0);
+  CUDADriver::get_instance().device_get_attribute(
+      &cc_major, CU_DEVICE_ATTRIBUTE_COMPUTE_CAPABILITY_MAJOR, device);
+  if (cc_major < 7) {
+    TI_WARN(
+        "CUPTI profiler APIs unsupported on Device with compute capability < "
+        "7.0 , fallback to default kernel profiler");
+    return false;
+  }
+  return true;
+}
+
+KernelProfilerCUDA::KernelProfilerCUDA() {
+// if Taichi was compiled with CUDA toolit, then use CUPTI
+// TODO : add set_mode() to select toolkit by user
+#if defined(TI_WITH_CUDA_TOOLKIT)
+  if (check_device_capability())
+    tool_ = ProfilingToolkit::cupti;
+#endif
+
+  if (tool_ == ProfilingToolkit::event) {
+    event_toolkit_ = std::make_unique<EventToolkit>();
+  } else if (tool_ == ProfilingToolkit::cupti) {
+    cupti_toolkit_ = std::make_unique<CuptiToolkit>();
+  }
+}
+
+// deprecated, move to trace()
+KernelProfilerBase::TaskHandle KernelProfilerCUDA::start_with_handle(
+    const std::string &kernel_name) {
+  TI_NOT_IMPLEMENTED;
+}
+
+void KernelProfilerCUDA::trace(KernelProfilerBase::TaskHandle &task_handle,
+                               const std::string &task_name) {
+  if (tool_ == ProfilingToolkit::event)
+    task_handle = event_toolkit_->start_with_handle(task_name);
+  else if (tool_ == ProfilingToolkit::cupti) {
+    TI_NOT_IMPLEMENTED;
+  }
+}
+
+void KernelProfilerCUDA::stop(KernelProfilerBase::TaskHandle handle) {
+  if (tool_ == ProfilingToolkit::event)
+    CUDADriver::get_instance().event_record(handle, 0);
+}
+
+bool KernelProfilerCUDA::statistics_on_traced_records() {
+  for (auto &record : traced_records_) {
+    auto it =
+        std::find_if(statistical_results_.begin(), statistical_results_.end(),
+                     [&](KernelProfileStatisticalResult &result) {
+                       return result.name == record.name;
+                     });
+    if (it == statistical_results_.end()) {
+      statistical_results_.emplace_back(record.name);
+      it = std::prev(statistical_results_.end());
+    }
+    it->insert_record(record.kernel_elapsed_time_in_ms);
+    total_time_ms_ += record.kernel_elapsed_time_in_ms;
+  }
+
+  return true;
+}
+
+void KernelProfilerCUDA::sync() {
+  // sync
+  CUDADriver::get_instance().stream_synchronize(nullptr);
+
+  // update
+  if (tool_ == ProfilingToolkit::event) {
+    event_toolkit_->update_record(traced_records_);
```

Comment:
wrap these into a EventToolkit::sync()?

Reply:
thanks! I'll get this done in the next PR today